crates/proof-of-sql/src/base/database/column_operation.rs

```diff
@@ -0,0 +1,364 @@
+            operator,
+            left_type: lhs,
+            right_type: rhs,
+        });
+    }
+    if lhs.is_integer() && rhs.is_integer() {
+        // We can unwrap here because we know that both types are integers
+        return Ok(lhs.max_integer_type(&rhs).unwrap());
+    }
+    if lhs == ColumnType::Scalar || rhs == ColumnType::Scalar {
+        Ok(ColumnType::Scalar)
+    } else {
+        let left_precision_value = lhs.precision_value().unwrap_or(0) as i16;
+        let right_precision_value = rhs.precision_value().unwrap_or(0) as i16;
+        let left_scale = lhs.scale().unwrap_or(0);
+        let right_scale = rhs.scale().unwrap_or(0);
+        let scale = left_scale.max(right_scale);
+        let precision_value: i16 = scale as i16
+            + (left_precision_value - left_scale as i16)
+                .max(right_precision_value - right_scale as i16)
+            + 1_i16;
+        let precision = u8::try_from(precision_value)
+            .map_err(|_| {
+                ColumnOperationError::DecimalConversionError(DecimalError::InvalidPrecision(
+                    precision_value.to_string(),
+                ))
+            })
+            .and_then(|p| {
+                Precision::new(p).map_err(|_| {
+                    ColumnOperationError::DecimalConversionError(DecimalError::InvalidPrecision(
+                        p.to_string(),
+                    ))
+                })
+            })?;
+        Ok(ColumnType::Decimal75(precision, scale))
+    }
+}
+
+/// Determine the output type of a multiplication operation if it is possible
+/// to multiply the two input types. If the types are not compatible, return
+/// an error.
+pub fn try_multiply_column_types(
+    lhs: ColumnType,
+    rhs: ColumnType,
+) -> ColumnOperationResult<ColumnType> {
+    if !lhs.is_numeric() || !rhs.is_numeric() {
+        return Err(ColumnOperationError::BinaryOperationInvalidColumnType {
+            operator: BinaryOperator::Multiply,
+            left_type: lhs,
+            right_type: rhs,
+        });
+    }
+    if lhs.is_integer() && rhs.is_integer() {
+        // We can unwrap here because we know that both types are integers
+        return Ok(lhs.max_integer_type(&rhs).unwrap());
+    }
+    if lhs == ColumnType::Scalar || rhs == ColumnType::Scalar {
+        Ok(ColumnType::Scalar)
+    } else {
+        let left_precision_value = lhs.precision_value().unwrap_or(0);
+        let right_precision_value = rhs.precision_value().unwrap_or(0);
+        let precision_value = left_precision_value + right_precision_value + 1;
+        let precision = Precision::new(precision_value).map_err(|_| {
+            ColumnOperationError::DecimalConversionError(DecimalError::InvalidPrecision(format!(
+                "Required precision {} is beyond what we can support",
+                precision_value
+            )))
+        })?;
+        let left_scale = lhs.scale().unwrap_or(0);
```

Comment:
If `scale` returns none, it should mean there is no scale (e.g. VarChar).
If the type actually has a scale of 0, then scale should return `Some(0)`.

Reply:
Same applies to `precision_value`.